apps/hashdotdev/src/pages/roadmap/technology-tree.tsx

```diff
@@ -0,0 +1,423 @@
+                      fontWeight: 400,
+                    }}
+                  >
+                    Please select at least one{" "}
+                    <strong>
+                      {displayedStatuses.length === 0
+                        ? "status"
+                        : displayedVariants.length === 0
+                        ? "type"
+                        : "use case"}
+                    </strong>{" "}
+                    to view matching deliverables
+                  </Typography>
+                ) : null}
+              </Box>
+            ) : null}
+            <Box
+              ref={graphWrapperRef}
+              sx={{
+                position: "relative",
+                width: "100%",
+                overflow: "hidden",
+                height: "50vh",
+                borderBottomColor: ({ palette }) => palette.gray[20],
+                borderBottomStyle: "solid",
+                borderBottomWidth: 1,
+              }}
+            >
+              <Box ref={graphRef}>
+                <Box
+                  sx={{
+                    position: "relative",
+                    zIndex: 1,
+                  }}
+                >
+                  {nodes.map(({ x, y, data }) => (
+                    <TechnologyTreeNode
+                      key={data.id}
+                      x={x}
+                      y={y}
+                      data={data}
+                      blurred={blurredNodes.includes(data.id)}
+                      onHover={() => setFocusedNodeId(data.id)}
+                      onUnhover={() => setFocusedNodeId(undefined)}
+                    />
+                  ))}
+                </Box>
+                <Box
+                  component="svg"
+                  sx={{
+                    position: "relative",
+                    width: layoutWidth + 4,
+                    height: layoutHeight + 4,
+                  }}
+                >
+                  <g transform="translate(2, 2)">
+                    <defs id="defs">
+                      <marker
+                        id="arrow-head"
+                        refX="6"
+                        refY="6"
+                        markerWidth="10"
+                        markerHeight="16"
+                        markerUnits="userSpaceOnUse"
+                        orient="auto-start-reverse"
+                      >
+                        <Box
+                          component="path"
+                          d="M 1 1 L 6 6 L 1 11.25"
+                          fill="none"
+                          stroke="blue"
+                          strokeWidth="1"
+                          sx={{ stroke: ({ palette }) => palette.gray[40] }}
+                        />
+                      </marker>
+                    </defs>
+                    <g id="links">
+                      {links
+                        /** Ensure there aren't any duplicate edges to avoid duplicate react keys */
```

Comment:
I'm not sure, it didn't make sense to me either but it was somehow occurring when selecting different filters so this was how I addressed it.

And yes this removes duplicate links, which prevents there from being duplicate react keys because those are derived from the `source` and `target` IDs of a link.

Reply:
I would update the comment to say that we are aiming to avoid duplicate link elements, instead of saying it's avoiding duplicate react keys – the way to fix duplicate React keys is to give each element a distinct key. But it's the element themselves we don't want

Reply:
The error happens to be surfaced by a React warning, but even if there were no such thing as React keys we still wouldn't want multiple of the same link sitting on top of each other

Reply:
Looks good, thank you